### PR TITLE
enforce a minimum square size

### PR DESCRIPTION
**Pull request by evan-forbes**
## Description

add the constant `MinSquareSize` and make sure that empty blocks always have a square size equal to it.

Notably, I wasn't sure if the hash of an empty data availability header should always be that of a data availability header of an empty block, which is not empty, so this is not currently enforced.

Closes: #281

### CODE REVIEW

File: types/shares_test.go [hunk 1]
```diff
@@ -446,10 +446,6 @@ func Test_parseDelimiter(t *testing.T) {
 	}
 }
 
-// ////////////////////////////
-// Test data generation
-// ////////////////////////////
```
Comment (evan-forbes) [inline types/shares_test.go:451]: cleaning up ascii graffiti

File: types/block.go [hunk 2]
```diff
@@ -1384,6 +1384,12 @@ func (data *Data) ComputeShares() (NamespacedShares, int) {
 	// FIXME(ismail): this is not a power of two
 	// see: https://github.com/lazyledger/lazyledger-specs/issues/80 and
 	wantLen := getNextSquareNum(curLen)
+
+	// ensure that the min square size is used
+	if wantLen < minSharecount {
+		wantLen = minSharecount
+	}
```
Comment (liamsi) [inline types/block.go:1391]: So that is the only change to ensure the data is padded to the minSharcount

**Review by liamsi — DISMISSED** [review]
👍🏼 Thanks for the quick fix.

**Comment by liamsi** [comment]
> Notably, I wasn't sure if the hash of an empty data availability header should always be that of a data availability header of an empty block, which is not empty, so this is not currently enforced.

Actually, I think this should be the case. cc @adlerjohn

**Comment by liamsi** [comment]
I still think we can go and merge this as is as it is an improvement to the status quo.

**Comment by evan-forbes** [comment]
I'm still trying to figure out why the CI is failing. everything passes locally... the e2e test isn't even able to start, so I'm guessing some bug with github actions or the docker container?

```
Run ./build/runner -f networks/ci.toml logs
  ./build/runner -f networks/ci.toml logs
  shell: /usr/bin/bash -e {0}
  env:
    GOROOT: /opt/hostedtoolcache/go/1.15.11/x64
    GIT_DIFF: 'types/block.go' 'types/block_test.go' 'types/consts.go' 'types/shares_test.go'
    GIT_DIFF_FILTERED: 'types/block.go' 'types/block_test.go' 'types/consts.go' 'types/shares_test.go'
    MATCHED_FILES: 
.FileNotFoundError: [Errno 2] No such file or directory: './networks/ci/docker-compose.yml'
E[2021-04-15|17:57:29.668] exit status 1                                
Error: Process completed with exit code 1.
```

**Comment by evan-forbes** [comment]
oh shoot docker is down https://downdetector.com/status/docker/

**Comment by evan-forbes** [comment]
interesting, using a minimum square size uncovered a data race in the codec map we're using in rsmt2d

<details>
<summary>
data race errro
</summary>
==================
WARNING: DATA RACE
Write at 0x00c0000f6030 by goroutine 275:
  runtime.mapassign_fast64()
      /usr/local/go/src/runtime/map_fast64.go:92 +0x0
  github.com/lazyledger/rsmt2d.(*rsGF8Codec).encode()
      /home/evan/go/pkg/mod/github.com/lazyledger/rsmt2d@v0.1.1-0.20210406153014-e1fd589bdb09/infectiousRSGF8.go:32 +0x28b
  github.com/lazyledger/rsmt2d.Encode()
      /home/evan/go/pkg/mod/github.com/lazyledger/rsmt2d@v0.1.1-0.20210406153014-e1fd589bdb09/codecs.go:53 +0x186
  github.com/lazyledger/rsmt2d.(*ExtendedDataSquare).erasureExtendSquare()
      /home/evan/go/pkg/mod/github.com/lazyledger/rsmt2d@v0.1.1-0.20210406153014-e1fd589bdb09/extendeddatasquare.go:85 +0x356
  github.com/lazyledger/rsmt2d.ComputeExtendedDataSquare()
      /home/evan/go/pkg/mod/github.com/lazyledger/rsmt2d@v0.1.1-0.20210406153014-e1fd589bdb09/extendeddatasquare.go:32 +0x1fb
  github.com/lazyledger/lazyledger-core/types.(*Block).fillDataAvailabilityHeader()
      /home/evan/go/src/github.com/lazyledger/lazyledger-core/types/block.go:241 +0x1ae
  github.com/lazyledger/lazyledger-core/types.(*Block).fillHeader()
      /home/evan/go/src/github.com/lazyledger/lazyledger-core/types/block.go:221 +0x24d
  github.com/lazyledger/lazyledger-core/types.MakeBlock()
      /home/evan/go/src/github.com/lazyledger/lazyledger-core/types/block.go:619 +0x28f
  github.com/lazyledger/lazyledger-core/state.State.MakeBlock()
      /home/evan/go/src/github.com/lazyledger/lazyledger-core/state/state.go:247 +0xfe
  github.com/lazyledger/lazyledger-core/state.(*BlockExecutor).CreateProposalBlock()
      /home/evan/go/src/github.com/lazyledger/lazyledger-core/state/execution.go:158 +0x62e
  github.com/lazyledger/lazyledger-core/consensus.(*State).createProposalBlock()
      /home/evan/go/src/github.com/lazyledger/lazyledger-core/consensus/state.go:1176 +0x5f7
  github.com/lazyledger/lazyledger-core/consensus.(*State).defaultDecideProposal()
      /home/evan/go/src/github.com/lazyledger/lazyledger-core/consensus/state.go:1076 +0x1314
  github.com/lazyledger/lazyledger-core/consensus.(*State).defaultDecideProposal-fm()
      /home/evan/go/src/github.com/lazyledger/lazyledger-core/consensus/state.go:1066 +0x53
  github.com/lazyledger/lazyledger-core/consensus.(*State).enterPropose()
      /home/evan/go/src/github.com/lazyledger/lazyledger-core/consensus/state.go:1052 +0x958
  github.com/lazyledger/lazyledger-core/consensus.(*State).enterNewRound()
      /home/evan/go/src/github.com/lazyledger/lazyledger-core/consensus/state.go:972 +0x95b
  github.com/lazyledger/lazyledger-core/consensus.(*State).addVote()
      /home/evan/go/src/github.com/lazyledger/lazyledger-core/consensus/state.go:2075 +0x1269
  github.com/lazyledger/lazyledger-core/consensus.(*State).tryAddVote()
      /home/evan/go/src/github.com/lazyledger/lazyledger-core/consensus/state.go:1871 +0x7c
  github.com/lazyledger/lazyledger-core/consensus.(*State).handleMsg()
      /home/evan/go/src/github.com/lazyledger/lazyledger-core/consensus/state.go:806 +0xd96
  github.com/lazyledger/lazyledger-core/consensus.(*State).receiveRoutine()
      /home/evan/go/src/github.com/lazyledger/lazyledger-core/consensus/state.go:755 +0x8c9

Previous read at 0x00c0000f6030 by goroutine 271:
  [failed to restore the stack]

Goroutine 275 (running) created at:
  github.com/lazyledger/lazyledger-core/consensus.(*State).startRoutines()
      /home/evan/go/src/github.com/lazyledger/lazyledger-core/consensus/state.go:399 +0x1f0
  github.com/lazyledger/lazyledger-core/consensus.startTestRound()
      /home/evan/go/src/github.com/lazyledger/lazyledger-core/consensus/common_test.go:183 +0x64
  github.com/lazyledger/lazyledger-core/consensus.TestMempoolTxConcurrentWithCommit()
      /home/evan/go/src/github.com/lazyledger/lazyledger-core/consensus/mempool_test.go:127 +0x693
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1123 +0x202

Goroutine 271 (running) created at:
  github.com/lazyledger/lazyledger-core/consensus.(*State).startRoutines()
      /home/evan/go/src/github.com/lazyledger/lazyledger-core/consensus/state.go:399 +0x1f0
  github.com/lazyledger/lazyledger-core/consensus.startTestRound()
      /home/evan/go/src/github.com/lazyledger/lazyledger-core/consensus/common_test.go:183 +0x64
  github.com/lazyledger/lazyledger-core/consensus.TestMempoolProgressAfterCreateEmptyBlocksInterval()
      /home/evan/go/src/github.com/lazyledger/lazyledger-core/consensus/mempool_test.go:58 +0x44c
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1123 +0x202
==================
--- FAIL: TestMempoolTxConcurrentWithCommit (2.45s)
    testing.go:1038: race detected during execution of test
FAIL
exit status 1
FAIL	github.com/lazyledger/lazyledger-core/consensus	26.799s

</details>

**Review by liamsi — APPROVED** [review]
🙌🏼 🤸‍♂️🤾‍♂️👑

**Pull request merged**